hypervisor/src/kvm/mod.rs

```diff
@@ -696,9 +696,35 @@ impl cpu::Vcpu for KvmVcpu {
         match self.fd.run() {
             Ok(run) => match run {
                 #[cfg(target_arch = "x86_64")]
-                VcpuExit::IoIn(addr, data) => Ok(cpu::VmExit::IoIn(addr, data)),
+                VcpuExit::IoIn(addr, data) => {
+                    if let Some(ref vmmops) = self.vmmops {
+                        let ret = vmmops.lock().unwrap().pio_read(addr.into(), data);
+                        if ret.is_ok() {
+                            return Ok(cpu::VmExit::Ignore);
+                        }
+                    }
+                    Err(cpu::HypervisorCpuError::RunVcpu(anyhow!(
+                        "Guest PIO read failed at address 0x{:x}",
+                        addr
+                    )))
+                }
                 #[cfg(target_arch = "x86_64")]
-                VcpuExit::IoOut(addr, data) => Ok(cpu::VmExit::IoOut(addr, data)),
+                VcpuExit::IoOut(addr, data) => {
+                    if let Some(ref vmmops) = self.vmmops {
+                        let ret = vmmops.lock().unwrap().pio_write(addr.into(), data);
+                        if ret.is_ok() {
+                            if addr == DEBUG_IOPORT && data.len() == 1 {
```

Comment:
The special handling of DEBUG_IOPORT should not be here. It should be in `pio_write`.

Reasons:
1. It is not specific to KVM.
2. It helps you reduce the amount of code churn -- you won't need to define DEBUG_IOPORT in `hypervisor` and at the same time introduce a `LogDebugPort`.

Reply:
log_debug_ioport was not available via vmmops and so I thought I could pass the control back to VmExit for invoking log_debug_ioport.

Thanks for clarifying DEBUG_IOPORT is specific to CH and not KVM. I will move it back into pio_write and figure out piping for log_debug_ioport.